### PR TITLE
[DROOLS-6908] Arithmetic operation with String coercion in constraint…

### PR DESCRIPTION
**Pull request by tkobayas**
**Ports** 
This PR is for 7.x
For main -> https://github.com/kiegroup/drools/pull/4332

**JIRA**: 
https://issues.redhat.com/browse/DROOLS-6908

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>a pull request</b> please add comment: <b>Jenkins retest this</b>

* for a <b>full downstream build</b> 
  * for <b>jenkins</b> job: please add comment: <b>Jenkins run fdb</b>
  * for <b>github actions</b> job: add the label `run_fdb`

* <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

* <b>a full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>

* <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>
</details>

### CODE REVIEW

File: drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/builder/generator/drlxparse/ArithmeticCoercedExpression.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        }
+
+        TypedExpression coercedLeft = left;
+        TypedExpression coercedRight = right;
+        if (leftClass == String.class) {
+            if (operator == Operator.PLUS) {
+                // String concatenation : Compatibility with mvel
+                coercedRight = coerceToString(right);
+            } else {
+                coercedLeft = coerceToDouble(left);
```
Comment (tkobayas) [inline drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/builder/generator/drlxparse/ArithmeticCoercedExpression.java:75]: We may coerce to `BigDecimal` but Mvel `MathProcessor` uses `double` so I followed the same.

File: drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/builder/generator/expressiontyper/ExpressionTyper.java [hunk 2]
```diff
@@ -212,10 +213,26 @@ private Optional<TypedExpression> toTypedExpressionRec(Expression drlxExpr) {
...
+            TypedExpression left = optLeft.get();
+            TypedExpression right = optRight.get();
+
+            ArithmeticCoercedExpression.ArithmeticCoercedExpressionResult coerced;
+            try {
+                coerced = new ArithmeticCoercedExpression(left, right, operator).coerce();
+            } catch (ArithmeticCoercedExpression.ArithmeticCoercedExpressionException e) {
+                logger.error("Failed to coerce : " + e.getInvalidExpressionErrorResult());
+                return empty();
+            }
```
Comment (tkobayas) [inline drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/builder/generator/expressiontyper/ExpressionTyper.java:229]: `CoercedExpression` is used in `ConstraintParser.parseBinaryExpr` for comparison (e.g. name == 40) at the end of parsing. But arithmetic operations happen during internal/recursive parsing so I placed this coercion in `ExpressionTyper.toTypedExpressionRec`.

File: drools-model/drools-mvel-compiler/src/main/java/org/drools/mvelcompiler/ast/ObjectCreationExpressionT.java [hunk 3]
```diff
@@ -45,7 +44,7 @@ public Optional<Type> getType() {
     @Override
     public Node toJavaExpression() {
         ObjectCreationExpr objectCreationExpr = new ObjectCreationExpr();
-        objectCreationExpr.setType(type);
+        objectCreationExpr.setType(type.getCanonicalName());
```
Comment (tkobayas) [inline drools-model/drools-mvel-compiler/src/main/java/org/drools/mvelcompiler/ast/ObjectCreationExpressionT.java:47]: Without this fix, a constructor is generated without class package (e.g. `new BigDecimal(...)`) thus causes a compile error.

File: drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/builder/generator/drlxparse/ArithmeticCoercedExpression.java (hunk 1 above)
Comment (mariofusco) [reply drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/builder/generator/drlxparse/ArithmeticCoercedExpression.java:75]: Please leave a comment here (or at least add it when you'll forward port this fix on main branch): I think that we cannot do anything different for 7.x, but at least for drools 8 we may want to drop mvel compatibility and do "the right thing".

**Review by mariofusco — APPROVED** [review]

File: drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/builder/generator/drlxparse/ArithmeticCoercedExpression.java (hunk 1 above)
Comment (tkobayas) [reply drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/builder/generator/drlxparse/ArithmeticCoercedExpression.java:75]: Sure, added a comment in the source.

**Review by hellowdan — APPROVED** [review]

**Comment by tkobayas** [comment]
@mariofusco  Please merge this PR, thanks!

**Comment by tkobayas** [comment]
Rebased because of conflict.

**Comment by tkobayas** [comment]
@mariofusco Please merge this PR, thanks!

**Pull request merged**